File: android/src/main/kotlin/com/anhlp/barcode_scanner/BoundingBoxOverlay.kt
package com.anhlp.barcode_scanner

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.graphics.Rect
import android.graphics.RectF
import android.view.View

class BoundingBoxOverlay(context: Context) : View(context) {
    private val path = Path()

    private val paint = Paint().apply {
        color = android.graphics.Color.GREEN
        style = Paint.Style.STROKE
        strokeWidth = 5f
    }

    private var rect: Rect? = null

    fun clear() {
        rect = null
        invalidate()
    }

    fun setBoundingBox(rectF: RectF, imageWidth: Int, imageHeight: Int) {
        val scaleX = width.toFloat() / imageWidth
        val scaleY = height.toFloat() / imageHeight

        val scale = maxOf(scaleX, scaleY)

        val wLost = (imageWidth * scale - width) / 2f
        val hLost = (imageHeight * scale - height) / 2f

        rect = Rect(
            (scale * rectF.left - wLost).toInt(),
            (scale * rectF.top - hLost).toInt(),
            (scale * rectF.right - wLost).toInt(),
            (scale * rectF.bottom - hLost).toInt()
        )

        invalidate()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        path.reset()
        rect?.let {canvas.drawRect(it, paint)}
    }
}

File: ios/Classes/BoundingBoxOverlay.swift
class BoundingBoxOverlay: UIView {
    private var rect: CGRect?
    
    public func clear() {
        self.rect = nil
        setNeedsDisplay()
    }
    
    public func setBoundingBox(_ rect: CGRect, imageSize: CGSize) {
        let imageViewSize = self.bounds.size
        
        let scaleX = imageViewSize.width / imageSize.width
        let scaleY = imageViewSize.height / imageSize.height
        let scale = min(scaleX, scaleY)
        
        let wLost = (scale == scaleY) ? (imageSize.width * scale - imageViewSize.width) / 2 : 0
        let hLost = (scale == scaleY) ? 0 : (imageSize.height * scale - imageViewSize.height) / 2
        
        let x = rect.origin.x * imageSize.width * scale - wLost
        let y = rect.origin.y * imageSize.height * scale - hLost
        let width = rect.width * imageSize.width * scale
        let height = rect.height * imageSize.height * scale
        
        self.rect = CGRect(x: x, y: y, width: width, height: height)
        setNeedsDisplay()
    }
    
    override func draw(_ rect: CGRect) {
        guard let boundingRect = self.rect else { return }
        
        let path = UIBezierPath(rect: boundingRect)
        path.lineWidth = 3.0
        UIColor.green.setStroke()
        path.stroke()
    }
}
